### PR TITLE
Add check for unknown attributes

### LINKED ISSUE #3994: DX: Helpful output when serverless.yml is malformed
# This is a Feature Proposal

## Description
If a user attempts to put a user-defined top-level property in serverless.yml (other than one called `custom`), it is silently ignored while packaging and anything referencing that property will fail down the line, potentially not until deployment time.

I think this issue is compounded by the fact that the documentation isn't entirely clear on the fact that a custom top-level property in the yml file **must** be called `custom`.

* What went wrong?
`serverless package` does not call out that an improperly named top-level property is defined and will be ignored.  Instead, there is only a warning that properties referencing this top-level property have no value.

* What did you expect should have happened?
  - If `custom` is the only name allowed for user-defined top level properties, I would expect a warning or error at the `serverless package` step to alert the user that what they're trying to do isn't going to work.
  - I'd also expect that the [documentation](https://serverless.com/framework/docs/providers/aws/guide/variables/) would have a sentence explicitly calling out that this naming restriction exists.

* What was the config you used?

Example serverless.yml exerpt:
```yaml
service: foo

# This will work
custom: someValue

# This will not
bar: someValue

provider:
...
```

## Feature proposals:
* `serverless package` prints a warning or error when it detects a top-level property which it's going to ignore.
* Documentation is improved to be more explicit about the `custom` naming convention.
  - If this is desired, I can look at putting in a PR.

## Additional Data

* ***Serverless Framework Version you're using***: 1.18

### PR DESCRIPTION
**Pull request by kbariotis**
## What did you implement:

Closes #3994

## How did you implement it:

Added a check against the known attributes and the current input.

## How can we verify it:

Add an unknown top level attribute in a serverless config and it will throw an error.

***Is this ready for review?:*** YES
***Is it a breaking change?:*** YES (since some plugins using the root of the serverless file to add extra information)

### CODE REVIEW

File: lib/classes/Service.js [hunk 1]
```diff
@@ -65,6 +76,14 @@ class Service {
       .parse(serviceFilePath)
       .then((serverlessFileParam) => {
         const serverlessFile = serverlessFileParam;
+
+        // check for non recognized attributes
+        Object.keys(serverlessFile).forEach((key) => {
+          if (this.allowedProperties.indexOf(key) === -1) {
+            throw new ServerlessError(`"${key}" is not recognized. If you are trying to use a custom attribute please use the "custom" object.`); // eslint-disable-line max-len
```
Comment (RafalWilinski) [inline lib/classes/Service.js]: One minor thing: I'd split that into multiple lines just like here: https://github.com/serverless/serverless/pull/3959/files#diff-1a6cb15a7aef359bf1fabed7982e4af3R107-L111

File: lib/classes/Service.test.js [hunk 2]
```diff
@@ -356,6 +356,24 @@ describe('Service', () => {
...
+      };
+
+      SUtils.writeFileSync(path.join(tmpDirPath, 'serverless.yaml'),
+        YAML.dump(serverlessYaml));
+
+      const serverless = new Serverless({ servicePath: tmpDirPath });
+      serviceInstance = new Service(serverless);
+
+      return expect(serviceInstance.load()).to.eventually.be
+        .rejectedWith('"nonExisting" is not recognized. If you are trying to use a custom attribute please use the "custom" object.'); // eslint-disable-line max-len
```
Comment (RafalWilinski) [inline lib/classes/Service.test.js]: Same here

**Comment by kbariotis** [comment]
@RafalWilinski @pmuens comments have been resolved. This is ready to be reviewed again. :)

**Comment by pmuens** [comment]
> @RafalWilinski @pmuens comments have been resolved. This is ready to be reviewed again. :)

Thanks for updating @kbariotis! 👍 We'll look into it soon!

**Comment by kbariotis** [comment]
@pmuens I do see your concerns about having to maintain another list of attributes although I would argue that `Service.js` would be the right place to put that list and also the code as is right now will complain in every unknown property so the developer will remember to update the list. 

But anyhow, which file are you referring to? I will check how else could that be done. Any other suggetion is ofc welcomed. :)

Thank you for the great work

**Comment by pmuens** [comment]
> @pmuens I do see your concerns about having to maintain another list of attributes although I would argue that Service.js would be the right place to put that list and also the code as is right now will complain in every unknown property so the developer will remember to update the list.

> But anyhow, which file are you referring to? I will check how else could that be done. Any other suggetion is ofc welcomed. :)

> Thank you for the great work

Thanks for the kind words and your feedback in this one @kbariotis 👍!

So generally speaking the whole parsing is taking place [here](https://github.com/serverless/serverless/blob/03afbcb03fef26c5665432f68a7ca03789f4657c/lib/classes/Service.js#L64-L122). While parsing the file we're basically filling out the [default properties](https://github.com/serverless/serverless/blob/03afbcb03fef26c5665432f68a7ca03789f4657c/lib/classes/Service.js#L13-L25) which already represent a list of all the supported properties a `serverless.yml` file can have.

So when we parse the `.yml` file we could basically check at the end if there are still properties in the files which we haven't already processed. If that's the case we could throw an error.

This way we don't have to introduce and maintain a separate list of supported properties, since they are already defined and filled out during the parsing process (so every time we add a new property and add support for it in the `parse` method it's automatically considered as a valid property when the code you introduce here checks for invalid properties).

Does that make sense to you? Let us know if you have any questions about that.

**Comment by kbariotis** [comment]
@pmuens I changed the PR. Smth I don't like about this approach is that the constructor of `Service` contains other properties as well and not just the defaults. Practically someone could pass a `serviceObject` property and still pass the validation. I thought I could move those into another property but the tests are [very coupled to its structure](https://github.com/serverless/serverless/blob/master/lib/classes/Service.test.js#L26) and I am not sure if anything else is also too.

If you like more this approach, awesome!

**Comment by kbariotis** [comment]
@pmuens @eahefnawy I have returned this to the previous commit where an array contains all the allowed attributes. Hope you are ok with this. 

We could also merge this and I could take the lead to refactor the Service class. What I am thinking is removing every outside reference to the inner properties of the class, by replacing them with public functions. Much like the `getAllFunctions` function.

Thank you :)

**Comment by HyperBrain** [comment]
@kbariotis One additional info about the Service class / object is, that lots of plugins store vital information there (or even modify properties of the contained service.service) because the service object is persisted from package to deploy and that way a plugin can seamlessly support package/deploy and they also need to affect the behavior of subsequent plugins.

A change/refactoring of the Service class is most likely breaking, and a test just with core and the core plugins is not sufficient to prove that it is not. If such a change is done, it should be at least verified, that it does not break the top 10 (15?) most popular serverless plugins. See here for a quite nice list:
https://www.serverlessconsultants.com/plugins/

**Comment by kbariotis** [comment]
@HyperBrain good to know. That would indeed be a breaking change. 👍

**Comment by pmuens** [comment]
Thanks for the update @kbariotis and the initiative to do a refactor 👍

Yes, @HyperBrain is right. Unfortunately we cannot touch that much code over there since many plugins depend on it. Maybe we can plan that for Serverless v2.

**Comment by horike37** [comment]
We are closing this PR for now, however, I just added it to the v2 milestone so that we can re-visit it once we're there.

**Pull request closed without merging**